rmf_site_editor/src/interaction/camera_controls/keyboard.rs

```diff
@@ -0,0 +1,365 @@
+};
+
+// Keyboard control limits
+pub const MIN_RESPONSE_TIME: f32 = 0.25; // [s] time taken to reach minimum input, or to reset
+pub const MAX_RESPONSE_TIME: f32 = 1.5; // [s] time taken to reach maximum input
+pub const MAX_FOV_DELTA: f32 = 0.5; // [rad/s]
+pub const MAX_ANGULAR_VEL: f32 = 3.5; // [rad/s]
+pub const MAX_TRANSLATION_VEL: f32 = 8.0; // [m/s]
+pub const SCALE_ZOOM_SENSITIVITY: f32 = 0.035;
+pub const ORTHOGRAPHIC_PAN_SENSITIVITY: f32 = 0.015;
+
+#[derive(Resource)]
+pub struct KeyboardCommand {
+    pub translation_delta: Vec3,
+    pub rotation_delta: Quat,
+    pub scale_delta: f32,
+    pub fov_delta: f32,
+    pub keyboard_motion: Vec2,
+    pub zoom_motion: f32,
+    pub camera_selection: Option<Vec3>,
+    pub command_type: CameraCommandType,
+}
+
+impl Default for KeyboardCommand {
+    fn default() -> Self {
+        Self {
+            translation_delta: Vec3::ZERO,
+            rotation_delta: Quat::IDENTITY,
+            scale_delta: 0.0,
+            fov_delta: 0.0,
+            keyboard_motion: Vec2::ZERO,
+            zoom_motion: 0.0,
+            camera_selection: None,
+            command_type: CameraCommandType::Inactive,
+        }
+    }
+}
+
+impl KeyboardCommand {
+    pub fn take_translation_delta(&mut self) -> Vec3 {
+        std::mem::replace(&mut self.translation_delta, Vec3::ZERO)
+    }
+
+    pub fn take_rotation_delta(&mut self) -> Quat {
+        std::mem::replace(&mut self.rotation_delta, Quat::IDENTITY)
+    }
+
+    pub fn take_scale_delta(&mut self) -> f32 {
+        std::mem::replace(&mut self.scale_delta, 0.0)
+    }
+
+    pub fn take_fov_delta(&mut self) -> f32 {
+        std::mem::replace(&mut self.fov_delta, 0.0)
+    }
+}
+
+pub fn update_keyboard_command(
+    mut camera_controls: ResMut<CameraControls>,
+    mut keyboard_command: ResMut<KeyboardCommand>,
+    keyboard_input: Res<Input<KeyCode>>,
+    cameras: Query<(&Projection, &Transform)>,
+    immediate_raycast: Raycast,
+    time: Res<Time>,
+    primary_windows: Query<&Window, With<PrimaryWindow>>,
+) {
+    if let Ok(_) = primary_windows.get_single() {
+        // User inputs
+        let is_shifting = keyboard_input.pressed(KeyCode::ShiftLeft)
+            || keyboard_input.pressed(KeyCode::ShiftRight);
+        let mut target_keyboard_motion = Vec2::ZERO;
+        if keyboard_input.pressed(KeyCode::Up) {
+            target_keyboard_motion.y += 1.0;
+        }
+        if keyboard_input.pressed(KeyCode::Left) {
+            target_keyboard_motion.x += -1.0;
+        }
+        if keyboard_input.pressed(KeyCode::Down) {
+            target_keyboard_motion.y += -1.0;
+        }
+        if keyboard_input.pressed(KeyCode::Right) {
+            target_keyboard_motion.x += 1.0;
+        }
+        if target_keyboard_motion.length() > 0.0 {
+            target_keyboard_motion = target_keyboard_motion.normalize();
+        }
+
+        let mut target_zoom_motion = 0.0;
+        if keyboard_input.pressed(KeyCode::N) {
```

Comment:
I think `M` and `N` are likely to be used for other hotkeys. I'd suggest changing this to `PgUp` and `PgDn`.

Reply:
Knit: Smaller keyboards may not have `PgUp` and `PgDn`.

Reply:
I think there's usually an equivalent shortcut, often using the `Fn` key.

In the long-term we should really have a hotkey register that lets people customize what keys are used for hotkeys, but in the meantime I think assigning the zoom in/out feature to keys that aren't always available is okay. If someone doesn't have `PgUp`/`PgDn` they will almost certainly still have some way to scroll, and they'll be able to use that for zooming.